src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -413,6 +476,71 @@ public static UIntPtr CompareExchange(ref UIntPtr location1, UIntPtr value, UInt
             return (UIntPtr)CompareExchange(ref Unsafe.As<UIntPtr, int>(ref location1), (int)value, (int)comparand);
 #endif
         }
+
+        /// <summary>Compares two instances of the specified type <typeparamref name="T"/> for equality and, if they are equal, replaces the first one.</summary>
+        /// <param name="location1">The destination, whose value is compared with <paramref name="comparand"/> and possibly replaced.</param>
+        /// <param name="value">The value that replaces the destination value if the comparison results in equality.</param>
+        /// <param name="comparand">The object that is compared to the value at <paramref name="location1"/>.</param>
+        /// <returns>The original value in <paramref name="location1"/>.</returns>
+        /// <exception cref="NullReferenceException">The address of <paramref name="location1"/> is a null pointer.</exception>
+        /// <exception cref="NotSupportedException">An unsupported <typeparamref name="T"/> is specified.</exception>
+        /// <typeparam name="T">
+        /// The type to be used for <paramref name="location1"/>, <paramref name="value"/>, and <paramref name="comparand"/>.
+        /// This type must be a reference type, an enum type (i.e. typeof(T).IsEnum is true), or a primitive type (i.e. typeof(T).IsPrimitive is true).
+        /// </typeparam>
+        [Intrinsic]
```

Comment:
I think this will require hardening expansion of the Exchange/CompareExchange in the JIT against unsupported `T` so that it only expands primitives and object refs and we fallback to throwing the NotSupportedException.

Reply:
The VM and the JIT only expands method marked as `[Intrinsic]`. I do not think any of the IL expansion is kicking in with the current version of the change.

Reply:
Yes, I'm not sure how I managed to delete that and I'll fix that. But with regards to hardening, is there something in particular you're seeing that looks wrong or you're just making a general comment? Regardless, I'll put back the attribute and fix the overload selection copy/paste error.

Reply:
I expect that you will hit this assert: https://github.com/dotnet/runtime/blob/db1f075c281b443a7de8b4523a2e1ea17dc39af6/src/coreclr/jit/importercalls.cpp#L4115 for the unsupported cases (we can wait for the CI to run to validate that it really the case). Changing that assert into `break;` similar to other cases should fix the problem.

Reply:
> I expect that you will hit this assert:
> 
> https://github.com/dotnet/runtime/blob/db1f075c281b443a7de8b4523a2e1ea17dc39af6/src/coreclr/jit/importercalls.cpp#L4115
> 
> for the unsupported cases (we can wait for the CI to run to validate that it really the case). Changing that assert into `break;` similar to other cases should fix the problem.

The `else if (!varTypeIsIntegral(retType))` should handle things there already.

Reply:
> This should get a test that verifies the exception for structs though.

I'd already added one:
https://github.com/dotnet/runtime/pull/104558/files#diff-b7ba854effe0b40e624638bc9ea70e50950841a776d09a3a1f36df3922f9efdfR397